Defining Spyfish Toolkit library structure in sftk directory.

Eatkin opened the pull request:
This PR aims to start defining some more rigid structure to the toolkit with the generalisation of a Spyfish Toolkit in the directory sftk. Adding this directory to `PYTHONPATH` or using `sys.path.insert` will allow the import of sftk and its modules.

Specific changes:
* `common.py` has been added to centralise any constants and the loading of `.env` so that `load_dotenv` will only be loaded once. This also allows any constants to be easily altered in a central location.
* `log_config.py` aims to centralise logging to `~/.sftk/logs` ensuring consistency across the usage of the `sftk` library. Running `from sftk import log_config` within each module allows the Python standard `logging` library to be used to log to a central location and allows easy modification of the configuration. Utility functions are included to adjust logging and to return the logfile location.
* `email_handler.py` provides a wrapper for the Python standard `imaplib` and `email` libraries and aims to provide a bridge to convert return values from those libraries into a more user-friendly format. The `Email` dataclass provides a simple structure for emails containing attributes such as `subject`, `sender`, `body` etc. The `EmailHandler` class is a singleton class meaning only one instance of the class may exist at a time. It provides various functions that make retrieving emails easier and can convert to the more usable `Email` dataclass.
* `s3_handler.py` is a singleton class with limited functionality, ensuring only one `s3_client` exists at one time. The `s3_client` may be accessed and used as standard via the `S3Handler.s3` attribute.
* `ingression.py` aims to begin defining consistent data ingression/ingestion processes with a `DataIngression` base class that may later be extended. One child class `EmailIngression` has been created which provides functionality to extract data from email bodies and returns a dictionary with keywords searched as the key and a list of dictionaries as the values representing the data retrieved.

KalindiFonda commented on sftk/ingression.py:
```diff
@@ -0,0 +1,116 @@
+    NOTE: Once an email is opened imap will mark it as read
+    This means if an email FAILS to be processed it will be marked as read and not picked up again
+    This is why summary logging is important for record keeping
+    The original email_to_sharepoint_copy.py script does not handle this
```
we should probably rename this file, because the copy at the end really threw both of us off. Or maybe we won't need it.

KalindiFonda commented on sftk/common.py:
```diff
@@ -0,0 +1,53 @@
+S3_SHAREPOINT_MOVIE_CSV = os.getenv("S3_SHAREPOINT_MOVIE_CSV")
+S3_SHAREPOINT_SURVEY_CSV = os.getenv("S3_SHAREPOINT_SURVEY_CSV")
+S3_SHAREPOINT_SITE_CSV = os.getenv("S3_SHAREPOINT_SITE_CSV")
+S3_SHAREPOINT_SPECIES_CSV = os.getenv("S3_SHAREPOINT_SPECIES_CSV")
```
Can you share the .env file sample - an .env file with all the entries but no values defined?

KalindiFonda commented on sftk/ingression.py:
```diff
@@ -0,0 +1,116 @@
+            dict | None: The extracted JSON data or None if no data was extracted
+        """
+        message = self.email_handler.get_email_content(id)
+        # Exception raised when the email is empty
```
you mention an exception, not sure where it is? is it on lines 52/53? 

```
else:
                    logging.warning(f"No data extracted from email with ID: {id}")
```

Eatkin replied:
Raised by the email handler, will update to make comment more clear!

KalindiFonda commented on sftk/ingression.py:
```diff
@@ -0,0 +1,116 @@
+            return None
+        body = message['body']
+        datum = self.email_handler.extract_json_from_body(body)
+        return datum
```
We can review the email management code again when/if we start using it.

KalindiFonda reviewed (commented):
Awesome, thanks! 

Looks good to me. I guess ✨ AI ✨ gave some suggestions too, the empty/defined raise seemed actionable.

You can add instructions to do into the readme.
"Adding this directory to PYTHONPATH or using sys.path.insert will allow the import of sftk and its modules."

Eatkin commented:
* Added sample .env file in `etc`, updated logging to contain constants `LOG_PATH` and `LOG_LEVEL` which can be used to retrieve our current logfile location and logging level.
* Added additional docstrings to ingression and clarified comment
* Updated README with instructions for adding library to path
* Added thread safety for singleton classes (shouldn't be necessary as they won't be used from within multithreading ideally but can't hurt)
